bootstrap practice settings gui

bradymiller opened the pull request:
work in progress so can get early feedback

mdsupport commented on templates/pharmacies/general_edit.html:
```diff
@@ -1,73 +1,78 @@
-<input type="hidden" name="id" value="{$pharmacy->id|escape:'html'}" />
-<input type="hidden" name="process" value="{$PROCESS|escape:'html'}" />
+    <input type="hidden" name="form_id" value="{$pharmacy->id|escape:'html'}" />
+	<div class="form-group">
```
Few suggestions : 

Although we have chosen not to use a framework, if we have to touch bulk of code in a script we should implement model-view approach to let PHP do its job.  For this script it would be simpler if you define in a php array how the fields are going to be displayed.  Then the `body` should be php blocks emitting html code.  If there are exceptions, let javascript/jquery do its stuff as bootstrap intends.
Specifically, consider a sample block -
```
<div class="form-group">
	<label for="address_line1" class="control-label col-sm-2">{xl t='Address'|escape:'html'}</label>
	<div class="col-sm-8">
	<input type="text" id="address_line1" name="address_line1" class="form-control" value="{$pharmacy->address-line1|escape:'html'}" onKeyDown="PreventIt(event)">
	</div>
</div>

```
This could be split as -
```
$pharmacy_group = array (
	'address_line1' => array ( 'disp' => 'address_line1', title => xl('Required', 'class'=>'keyprevent', 'value'=>$pharmacy->address-line1) ),
	....

foreach($pharmacy_group as $fld_db => $fld_attr) {
....

	printf('
<div class="form-group">
	<label for="%s" class="control-label col-sm-2">%s</label>
	<div class="col-sm-8">
	<input type="text" id="%s" name="%s" class="form-control" value="%s">
	</div>
</div>',
$fld_attr['disp'], ....
...

$('.prevent').on('keypress', function() {
...
});
```

mdsupport replied:
Please consider the objective of layout.  If it is intended to be tabular, bootstrap supports col-* for tables if `<table>` includes class `table`.  Just add it to header or first row.

On the other hand, current code has strange uses of tables purely for layout.  It makes sense there to spend effort to convert to grid or any other form of layout.

bradymiller replied:
@mdsupport , I thought I removed the table here, though, and converted it to grid. Let me know if'm missing something.

bradymiller replied:
@mdsupport , The goal here was to migrate interface from using `table` to bootstrap grid. Although I'm touching a bulk of the code, my changes are superficial and I decided not to go too deep into the code (since there are literally hundreds of scripts to convert to bootstrap) :)
Also, in this case, trying to keep it pure smarty template to allow migration to smarty3 in the future.

mdsupport replied:
I was questioning value of spending development effort of replacing `table` by `div`.  If you were to write completely new code in responsive world for large number of devices, it makes sense to start with div.  But in cases like this, reworking the code manually is a luxury for the sake of purity when merely adding table class to existing code and stripping all td formatting let's bootstrap.js do the same rendering.

robertdown replied:
There is a ton of value in replacing `table`-based design to `css`. Currently table-based design is preventing us from creating cohesive and modern user interfaces and experiences. There's a lot of work happening on the UI/UX side of the code; Table-based design has not been considered acceptable practice in over a decade. As it stands, we are already working on rewriting the entire UI. It just has to be done over several steps, one of the first being the removal of all table-based design.

sunsetsystems replied:
I'd rather see the pros and cons of tables discussed in detail for specific cases. It's hard to see what the big deal is when speaking in generalities. Might be instructive to pick a script and do it both ways, and then compare the merits.

robertdown replied:
From the HTML spec:

> Tables should not be used as layout aids. Historically, many Web authors have tables in HTML as a way to control their page layout making it difficult to extract tabular data from such documents. In particular, users of accessibility tools, like screen readers, are likely to find it very difficult to navigate pages with tables used for layout. If a table is to be used for layout it must be marked with the attribute role="presentation" for a user agent to properly represent the table to an assistive technology and to properly convey the intent of the author to tools that wish to extract tabular data from the document. There are a variety of alternatives to using HTML tables for layout, primarily using CSS positioning and the CSS table model. [CSS]
https://www.w3.org/TR/html5/tabular-data.html#tabular-data

From only the above, the burden of defending the layout should fall squarely on the use of `table`, not HTML/CSS. Is there a time where the use of a table should be used to layout non-tabular data? Maybe, but it's the exception to the rule.

As I've stated before, table-based layouts have not been acceptable practice for near a decade, if not longer. They hinder modern development. The standing rule should be "Use `divs` and CSS unless using tabular data."

This should not decided on a case-by-case basis. We should use modern development standard.

We've for far too long been lax about standards, both in _how_ we write code (read: code styling) and _what_ we write (read: table-based layouts). Finally getting PSR in place has been a huge stride for the project, and dropping these tables will be another huge step.

sunsetsystems replied:
Would still like to see examples. We're throwing around expressions like "table-based layout", "table-based design" and "tabular data" without being clear about what the differences are. "Tabular data" in a table is OK? What determines if it's in that category?

Also I don't get why it's hard for a screen reader to parse tabular data from a table. I've done scraping like that before and tables were not a problem.

bradymiller replied:
Hi @sunsetsystems , @robertdown , @mdsupport ,

Would rather stay practical. This specific form I worked on was really straightforward to convert from using table to bootstrap grid. This then makes it clear how it is responsive (when screen is small the labels go over the text fields).

Check out my recent globals PR to get a flavor of a less invasive way to do where are basically substituting table->grid tr->row td->div
And note this can support normal form responsive behavior (look at the Administration->Globals) or a form that is tabular (requires several columns of input) which is illustrated in the User Settings use of this script.

I hope we aren't having to defend against just leaving these forms in their current table/tr/td form which is not responsive and doesn't leverage bootstrap.

-brady

bradymiller replied:
also to define Tabular Data,
I think the report outputs are perfect examples of this

mdsupport replied:
Should have been clearer in original comment that emphasis was on 'value of spending development effort'.  @robertdown , without in any way discounting the efforts of anybody, I would strongly disagree that _ton of value_ is going to be delivered by converting one layout structure to another. 
  Even more so if user continues to see the same tabular representation of data.  Taking this case as an example, I would rather have time and effort spent on creating a class or function that can render a simple recordset may be using datatables.

bradymiller replied:
@mdsupport , It can be tough to redirect volunteer resources (including myself :) ). In this case, these mods improved the codebase/gui(agree not to a huge degree) while also helping me to learn more about bootstrap and smarty.

sunsetsystems commented:
Thanks Brady, that's helpful. Responsive is good.

bradymiller commented on modules/sms_email_reminder/cron_sms_notification.php:
```diff
@@ -49,9 +49,9 @@
 $SMS_GATEWAY_PASSWORD = $vectNotificationSettings['SMS_gateway_password'];
 $SMS_GATEWAY_APIKEY = $vectNotificationSettings['SMS_gateway_apikey'];
-// set cron time (time to event ?) - todo extra tests
+// set cron time (time to event ?) - 1todo extra tests
```
oops, will revert this file changes since unrelated.

bradymiller commented:
Just brought in final code for this which is testing well. Basically converted the Administration->Practice to responsive bootstrap. Also cleaned up some stuff (for example, removed the sole remaining native php tag).

bradymiller commented:
testing very well so will bring in.
@sjpadgett , Note this updates the form code for the X12 Partners, so if you plan to edit this, rec bringing this code in first.

sjpadgett commented:
No problem, mine are a few edits to insurance smarty if I bring that part in at all. Finished x12 and am getting files dispersed for PR. Thanks for heads up.
